immuni_exposure_ingestion/models/validators.py

```diff
@@ -7,58 +9,152 @@
+_ROLLING_PERIOD_MIN = 0
+_ROLLING_PERIOD_MAX = 144
+
 
 class TekListValidator(Validator):
     """
     A validator for a list of TEKs.
     """
 
-    def __call__(self, value: List[TemporaryExposureKey]) -> None:
+    def __call__(self, teks: List[TemporaryExposureKey]) -> None:
         """
-        Validations to be performed according to Apple's documentation:
+        Perform the validation of a given list of TEKs, both on punctual value and as a whole.
 
-        - Any ENIntervalNumber values from the same user are not unique.
-        - The period of time covered by the data file exceeds 14 days.
-        - Any keys in the file have overlapping time windows.
-        - The TEKRollingPeriod value is not 144.
+        NOTE: The validation of a single TEKs could be performed at field level.
+          We expect (some of) these validations to change in the future. Having them all together
+          here aims at easing any future change in the validation steps.
 
-        :param value: the value of the schema being validated.
+        :param teks: the list of TEKs to be validated.
         """
 
-        if len(value) == 0:
-            return
-
-        # NOTE: This kind of validation (rolling_period == 144) could be performed at field level.
-        #   We expect all of these validations to change in the future, so having them all together
-        #   here is done on purpose to make things simpler in case of changes in validation.
-        #   We would like to keep field validation more flexible as not to spread strict validation
-        #   logic in many places.
+        n_teks = len(teks)
+        _LOGGER.info("Validating TEKs.", extra=dict(n_teks=n_teks))
 
-        if any(tek.rolling_period != 144 for tek in value):
-            raise ValidationError("Some rolling values are not 144.")
+        if n_teks == 0:
+            return
 
-        if (n_keys := len(value)) > config.MAX_KEYS_PER_UPLOAD:
+        if n_teks > config.MAX_KEYS_PER_UPLOAD:
             raise ValidationError(
-                f"Too many TEKs. (actual: {n_keys}, max_allowed: {config.MAX_KEYS_PER_UPLOAD})."
+                f"Too many TEKs. (actual: {n_teks}, max_allowed: {config.MAX_KEYS_PER_UPLOAD})."
             )
 
-        sorted_teks = sorted(value, key=operator.attrgetter("rolling_start_number"))
+        valid_teks = self._filter_valid_teks(teks)
+
+        sorted_teks = sorted(
+            valid_teks, key=operator.attrgetter("rolling_start_number")
+        )
         rolling_start_numbers = [t.rolling_start_number for t in sorted_teks]
 
+        # TEKs with the same rolling_start_number can exist when considering a rolling_period < 144.
+        # Though, since the latter are ignored, we check the rolling_start_number is unique.
         if len(rolling_start_numbers) != len(set(rolling_start_numbers)):
             raise ValidationError("Rolling start numbers are not unique.")
 
         next_rolling_start_number = sorted_teks[0].rolling_start_number
```

Comment:
as you are filtering out valid teks some lines above, can it happen that the sorted_teks list is empty, thus raising an exception here? :)

Reply:
True! Very unlikely but possible. Changing. 🙏